### PR TITLE
Create a base trait for row level operations - Foundational work for MERGE INTO

### PR DESCRIPTION
**Pull request by dilipbiswal**
- Creates a base trait ```RewriteRowLevelOperationHelper``` that all the rewrite rules for row level operations will make use of.
- Its a basic refactor PR which refactors the method from `RewriteDelete` rule.

### CODE REVIEW

**Comment by dilipbiswal** [comment]
cc @aokolnychyi

File: spark3-extensions/src/main/scala/org/apache/spark/sql/utils/RewriteRowLevelOperationHelper.scala [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+trait RewriteRowLevelOperationHelper extends PredicateHelper with Logging {
+  protected val FILE_NAME_COL = "_file"
+  protected val ROW_POS_COL = "_pos"
+
+  protected def buildScanPlan(table: Table,
+                    output: Seq[AttributeReference],
+                    mergeBuilder: MergeBuilder,
+                    cond: Expression): LogicalPlan = {
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/utils/RewriteRowLevelOperationHelper.scala:43]: Nit: indentation doesn't align.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDelete.scala [hunk 2]
```diff
@@ -87,35 +68,8 @@ object RewriteDelete extends Rule[LogicalPlan] with PredicateHelper with Logging
...
-        scanRelation
-    }
-  }
-
-  private def buildWritePlan(
-      remainingRowsPlan: LogicalPlan,
-      output: Seq[AttributeReference]): LogicalPlan = {
-
+  private def buildWritePlan(remainingRowsPlan: LogicalPlan,
+                             output: Seq[AttributeReference]): LogicalPlan = {
```
Comment (aokolnychyi) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDelete.scala:72]: Are these cosmetic changes needed?

Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDelete.scala:72]: Nit: this changes the indentation of the arguments and make it appear that this method was modified, when it wasn't. Could you revert the changes to `buildWritePlan`?

File: spark3-extensions/src/main/scala/org/apache/spark/sql/utils/RewriteRowLevelOperationHelper.scala (hunk 1 above)
Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/utils/RewriteRowLevelOperationHelper.scala:43]: yeah, I'd also put `table` on a new line:

```
  protected def buildScanPlan(
      table: Table,
      output: Seq[AttributeReference],
      mergeBuilder: MergeBuilder,
      cond: Expression): LogicalPlan = {
```

Args should be formatted with 4 spaces.

**Review by rdblue — CHANGES_REQUESTED** [review]
I think that this only moves code. If so, it looks functionally fine to me. There are style issues that I think should be fixed before merging though. The reformatting of `buildWritePlan` arguments will cause commit conflicts.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/utils/RewriteRowLevelOperationHelper.scala (hunk 1 above)
Comment (rdblue) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/utils/RewriteRowLevelOperationHelper.scala:43]: Yeah, we should eventually make sure our style checker catches these.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/utils/RewriteRowLevelOperationHelper.scala [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    val resolver = SQLConf.get.resolver
+    plan.output.find(attr => resolver(attr.name, attrName)).getOrElse {
+      throw new AnalysisException(s"Cannot find $attrName in ${plan.output}")
+    }
+  }
+
+  protected def toOutputAttrs(schema: StructType, output: Seq[AttributeReference]): Seq[AttributeReference] = {
+    val nameToAttr = output.map(_.name).zip(output).toMap
+    schema.toAttributes.map {
+      a =>
```
Comment (aokolnychyi) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/utils/RewriteRowLevelOperationHelper.scala:96]: nit: I'd prefer the original formatting where `nameToAttr.get(a.name)` is on the same line:

```
      a => nameToAttr.get(a.name) match {
        case Some(ref) =>
          // keep the attribute id if it was present in the relation
          a.withExprId(ref.exprId)
        case _ =>
          // if the field is new, create a new attribute
          AttributeReference(a.name, a.dataType, a.nullable, a.metadata)()
      }
```

Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/utils/RewriteRowLevelOperationHelper.scala:96]: Or putting a => on the line above like:

```
    schema.toAttributes.map { a =>
      nameToAttr.get(a.name) match {
        case Some(ref) =>
          // keep the attribute id if it was present in the relation
          a.withExprId(ref.exprId)
        case _ =>
          // if the field is new, create a new attribute
          AttributeReference(a.name, a.dataType, a.nullable, a.metadata)()
      }
    }
```

File: spark3-extensions/src/main/scala/org/apache/spark/sql/utils/RewriteRowLevelOperationHelper.scala [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.spark.sql.catalyst.utils
+
+import java.util.UUID
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.{sources, AnalysisException}
+import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeReference, Expression, PredicateHelper}
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, DynamicFileFilter, Filter, LogicalPlan, Project}
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/utils/RewriteRowLevelOperationHelper.scala:25]: Another nit: multiple imports on the same line cause lots of commit conflicts. Could you put these on separate lines?

**Comment by aokolnychyi** [comment]
Looks good to me apart from already pointed things.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/utils/RewriteRowLevelOperationHelper.scala [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+  private def buildFileFilterPlan(cond: Expression, scanRelation: DataSourceV2ScanRelation): LogicalPlan = {
+    val matchingFilter = Filter(cond, scanRelation)
+    val fileAttr = findOutputAttr(matchingFilter, FILE_NAME_COL)
+    val agg = Aggregate(Seq(fileAttr), Seq(fileAttr), matchingFilter)
+    Project(Seq(findOutputAttr(agg, FILE_NAME_COL)), agg)
+  }
+
+  protected def findOutputAttr(plan: LogicalPlan, attrName: String): Attribute = {
+    val resolver = SQLConf.get.resolver
```
Comment (rdblue) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/utils/RewriteRowLevelOperationHelper.scala:87]: Can this trait define `def resolver: Resolver` so it doesn't need to use `SQLConf.get`? The resolver should be provided by the rule this is mixed into.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/utils/RewriteRowLevelOperationHelper.scala [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    val normalizedPredicates = DataSourceStrategy.normalizeExprs(predicates, output)
+    PushDownUtils.pushFilters(scanBuilder, normalizedPredicates)
+
+    val scan = scanBuilder.build()
+    val scanRelation = DataSourceV2ScanRelation(table, scan, toOutputAttrs(scan.readSchema(), output))
+
+    scan match {
+      case filterable: SupportsFileFilter =>
+        val matchingFilePlan = buildFileFilterPlan(cond, scanRelation)
+        val dynamicFileFilter = DynamicFileFilter(scanRelation, matchingFilePlan, filterable)
```
Comment (aokolnychyi) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/utils/RewriteRowLevelOperationHelper.scala:57]: @dilipbiswal, can we get rid of `dynamicFileFilter` too? I know it is in the original code, I missed to update it.

Comment (aokolnychyi) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/utils/RewriteRowLevelOperationHelper.scala:57]: Let's just return `DynamicFileFilter(scanRelation, matchingFilePlan, filterable)` from this branch.

**Review by rdblue — APPROVED** [review]

**Review by aokolnychyi — APPROVED** [review]

**Comment by aokolnychyi** [comment]
Thanks, @dilipbiswal!
